File: backend/auth-node/src/services/authService.js
import bcrypt from 'bcrypt';
import { User, Role } from '../models/User.js';
import { JwtService } from './jwtService.js';

export class AuthService {
  static async registerUser(registerData) {
    const { username, email, password, firstName, lastName } = registerData;

    // Проверка существования пользователя
    if (await User.existsByUsername(username)) {
      throw new Error('Пользователь с таким именем уже существует');
    }

    if (await User.existsByEmail(email)) {
      throw new Error('Пользователь с таким email уже существует');
    }

    // Хеширование пароля
    const hashedPassword = await bcrypt.hash(password, 10);

    // Создание пользователя
    const user = await User.create({
      username,
      email,
      password: hashedPassword,
      firstName,
      lastName,
      role: Role.USER
    });

    return {
      message: 'Регистрация прошла успешно! Теперь вы можете войти в систему.',
      username: user.username
    };
  }

  static async loginUser(loginData) {
    const { username, password } = loginData;

    // Поиск пользователя
    const user = await User.findByUsername(username);
    if (!user) {
      throw new Error('Пользователь не найден');
    }

    // Проверка пароля
    const isPasswordValid = await bcrypt.compare(password, user.password);
    if (!isPasswordValid) {
      throw new Error('Неверный пароль');
    }

    // Проверка активности
    if (!user.active) {
      throw new Error('Пользователь в оффлайне');
    }

    // Генерация токена
    const token = JwtService.generateToken(user);

    return {
      message: 'Вход выполнен успешно',
      token,
      username: user.username
    };
  }
}

File: $RECYCLE.BIN/S-1-5-21-468433717-555027351-2740982991-1002/$RQXU4IE/src/main/java/com/_x/TestRegister1/Service/AuthService.java
package com._x.TestRegister1.Service;

import com._x.TestRegister1.DTO.LoginRequest;
import com._x.TestRegister1.DTO.RegisterRequest;
import com._x.TestRegister1.DTO.ResponseDto;
import com._x.TestRegister1.Jwt.JwtService;
import com._x.TestRegister1.Models.Role;
import com._x.TestRegister1.Models.User;
import com._x.TestRegister1.Repository.UserRepository;
import jakarta.transaction.Transactional;
import org.springframework.context.annotation.Bean;
import org.springframework.security.core.userdetails.UserDetails;
import org.springframework.security.crypto.bcrypt.BCryptPasswordEncoder;
import org.springframework.security.crypto.password.PasswordEncoder;
import org.springframework.stereotype.Service;

import java.time.LocalDateTime;

@Service
@Transactional
public class AuthService {

    private final UserRepository userRepository;
    private final PasswordEncoder passwordEncoder;
    private final JwtService jwtService;
    private final CustomUserDetailsService userDetailsService;

    public AuthService(UserRepository userRepository, PasswordEncoder passwordEncoder, JwtService jwtService, CustomUserDetailsService userDetailsService) {
        this.userRepository = userRepository;
        this.passwordEncoder = passwordEncoder;
        this.jwtService = jwtService;
        this.userDetailsService = userDetailsService;
    }


    public ResponseDto registerUser(RegisterRequest request) {
        if (userRepository.existsByUsername(request.getUsername())) {
            throw new RuntimeException("Пользователь с таким именем уже существует");
        }

        if (userRepository.existsByEmail(request.getEmail())) {
            throw new RuntimeException("Пользователь с таким email уже существует");
        }

        User user = new User();
        user.setUsername(request.getUsername());
        user.setEmail(request.getEmail());
        user.setPassword(passwordEncoder.encode(request.getPassword()));
        user.setFirstName(request.getFirstName());
        user.setLastName(request.getLastName());
        user.setCreatedAt(LocalDateTime.now());
        user.setActive(true);
        user.setRole(Role.USER);

        userRepository.save(user);


        return new ResponseDto(
                "Регистрация прошла успешно! Теперь вы можете войти в систему.",
                user.getUsername()
        );
    }


    public ResponseDto loginUser(LoginRequest request) {

        User user = userRepository.findByUsername(request.getUsername())
                .orElseThrow(() -> new RuntimeException("Пользователь не найден"));

        if (!passwordEncoder.matches(request.getPassword(), user.getPassword())) {
            throw new RuntimeException("Неверный пароль");
        }

        if (!user.isActive()) {
            throw new RuntimeException("Пользователь в оффлайне");
        }

        UserDetails userDetails = userDetailsService.loadUserByUsername(request.getUsername());
        String token = jwtService.generateToken(userDetails);

        return new ResponseDto(
                "Вход выполнен успешно",
                user.getUsername(),
                token
        );
    }
}

